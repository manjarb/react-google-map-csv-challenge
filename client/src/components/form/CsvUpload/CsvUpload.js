import React from 'react'
import { Formik } from 'formik'
import { connect } from 'react-redux'
import styled from 'styled-components'

import { uploadCsv } from '../../../redux/actions/index'

const Container = styled.div`
  padding-left: 15px;
  padding-right: 15px;
`

export class CsvUploadImp extends React.Component {
  formSubmit = (values, actions) => {
    const formData = new FormData()
    formData.append('fileUpload', values.fileUpload)
    formData.append('fileName', values.fileName)

    this.props.uploadCsv(formData)
    actions.resetForm({
      fileUpload: null,
      fileName: '',
    })
    actions.setSubmitting(false)
    this.fileUploadInput.value = null
  }

  formValidate = (values) => {
    const requiredText = 'Required'

    const errors = {}
    if (!values.fileUpload) {
      errors.fileUpload = requiredText
    }

    if (!values.fileName) {
      errors.fileName = requiredText
    }

    return errors
  }

  returnDangerClass = status => (status ? 'is-danger' : '')

  render() {
    return (
      <Formik
        initialValues={{
          fileUpload: null,
          fileName: '',
        }}
        validate={this.formValidate}
        onSubmit={this.formSubmit}
      >
        {({
          values,
          errors,
          touched,
          handleChange,
          handleSubmit,
          isSubmitting,
          setFieldValue,
        }) => (
          <Container>
            <form
              onSubmit={handleSubmit}
              data-testid="csv-upload-form-container"
            >
              <div className="columns">
                <div className="column is-6">
                  <div className="field">
                    <label
                      className="label"
                      htmlFor="portal-upload-form-file"
                    >
                      CSV file
                    </label>
                    <div className="control">
                      <input
                        id="portal-upload-form-file"
                        type="file"
                        name="fileUpload"
                        ref={(ref) => { this.fileUploadInput = ref }}
                        onChange={(event) => {
                          setFieldValue('fileUpload', event.currentTarget.files[0])
                        }}
                      />
                      <p
                        className="help is-danger"
                        data-testid="portal-upload-file-error"
                      >
                        {(errors.fileUpload && touched.fileUpload) && errors.fileUpload}
                      </p>
                    </div>
                  </div>
                </div>
                <div className="column is-6">
                  <div className="field">
                    <label
                      className="label"
                      htmlFor="portal-upload-form-name"
                    >
                      File name (No extension needed)
                    </label>
                    <div className="control">
                      <input
                        className={`input ${this.returnDangerClass(errors.fileName && touched.fileName)}`}
                        id="csv-upload-form-name"
                        type="text"
                        name="fileName"
                        onChange={handleChange}
                        value={values.fileName}
                        placeholder="File name"
                      />

                      <p
                        className="help is-danger"
                        data-testid="portal-upload-name-error"
                      >
                        {(errors.fileName && touched.fileName) && errors.fileName}
                      </p>
                    </div>
                  </div>
                </div>
              </div>
              <div className="submit-box">
                <button
                  className="button is-primary"
                  type="submit"
                  data-testid="csv-upload-submit-button"
                  disabled={isSubmitting}
                >
                  Upload CSV
                </button>
              </div>
            </form>
          </Container>
        )}
      </Formik>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  uploadCsv: payload => dispatch(uploadCsv(payload)),
})

export const CsvUpload = connect(null, mapDispatchToProps)(CsvUploadImp)
