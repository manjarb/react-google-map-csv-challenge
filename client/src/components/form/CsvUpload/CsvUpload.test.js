import React from 'react'
import {
  render,
  fireEvent,
  wait,
} from '@testing-library/react'
import { CsvUploadImp } from './CsvUpload'

describe('<CsvUploadImp />', () => {
  let testLabel
  let testText
  let testId
  let testDebug

  beforeEach(() => {
    const {
      getByLabelText, getByText, getByTestId, debug,
    } = render(
      <CsvUploadImp />,
    )

    testLabel = getByLabelText
    testText = getByText
    testId = getByTestId
    testDebug = debug
  })

  test('renders a form with file, name, uploaded by and a submit button', () => {
    expect(testId('csv-upload-form-container')).toHaveFormValues({
      fileUpload: '',
      fileName: '',
    })
  })

  test('renders an error validation', async () => {
    const fileError = testId('csv-upload-file-error')
    const nameError = testId('csv-upload-name-error')
    expect(fileError).toBeEmpty()
    expect(nameError).toBeEmpty()
    const submitButton = testId('csv-upload-submit-button')
    fireEvent.click(submitButton)
    await wait(() => {
      expect(fileError).toHaveTextContent('Required')
      expect(nameError).toHaveTextContent('Required')
    })
  })

  test('Snapshot matched', () => {
    expect(testId('csv-upload-form-container')).toMatchSnapshot()
  })
})
