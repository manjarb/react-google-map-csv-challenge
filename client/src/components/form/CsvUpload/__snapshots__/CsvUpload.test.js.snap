// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CsvUploadImp /> Snapshot matched 1`] = `
<form
  data-testid="csv-upload-form-container"
>
  <div
    class="columns"
  >
    <div
      class="column is-6"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="csv-upload-form-file"
        >
          CSV file
        </label>
        <div
          class="control"
        >
          <input
            accept="text/csv"
            id="csv-upload-form-file"
            name="fileUpload"
            type="file"
          />
          <p
            class="help is-danger"
            data-testid="csv-upload-file-error"
          />
        </div>
      </div>
    </div>
    <div
      class="column is-6"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="csv-upload-form-name"
        >
          File name (No extension needed)
        </label>
        <div
          class="control"
        >
          <input
            class="input "
            id="csv-upload-form-name"
            name="fileName"
            placeholder="File name"
            type="text"
            value=""
          />
          <p
            class="help is-danger"
            data-testid="csv-upload-name-error"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="submit-box"
  >
    <button
      class="button is-primary"
      data-testid="csv-upload-submit-button"
      type="submit"
    >
      Upload CSV
    </button>
  </div>
</form>
`;
